File: scripts/topology_adapter.py
# SPDX-License-Identifier: Apache-2.0
"""Independent offline A/B/C orchestration and observation adapter."""
import json
import os
import sys
import tempfile
import time
from copy import deepcopy
from pathlib import Path
from llm_fixtures import Fixture
from psp_cdl_llmproxy import BufferedLlmLoop, prompt_context
from psp_cdl_mcpproxy import McpDispatchGate, binding_digest
from psp_cdl_mcpproxy.mcp import StdioMcpClient
from psp_cdl_core.crypto import sign_envelope
from psp_cdl_cdl import tokenize_declaration
from topology_common import SUITE, CREDENTIAL, approval, environment, read_events, close_approval


class DirectError(Exception):
    def __init__(self, code): self.code = code


def run_direct(provider, registrations, message, options):
    """No PSP/CDL decision point. A host-owned provider callback drives this loop."""
    def check():
        if options['cancelled'](): raise DirectError('CANCELLED')
    messages = [{'role':'system','content':'Use the synthetic read tool when needed.'},{'role':'user','content':message}]
    tools = [{'name':r['server']+'.'+r['name'],'inputSchema':r['inputSchema'],'outputSchema':r['outputSchema']} for r in registrations]
    for _ in range(options['maxSteps']):
        check()
        result = provider['invoke'](deepcopy({'messages':messages,'tools':tools}),options)
        check()
        if type(result) is not dict: raise DirectError('INVALID_RESPONSE')
        if result.get('type') == 'final' and isinstance(result.get('text'),str): return result['text']
        tool = next((r for r in registrations if r['server']+'.'+r['name'] == result.get('name')),None)
        if result.get('type') != 'tool' or tool is None or type(result.get('arguments')) is not dict: raise DirectError('INVALID_RESPONSE')
        try: data = tool['invoke'](result['arguments'],options)
        except Exception:
            check()
            raise
        check()
        messages.extend([{'role':'assistant','call':{'name':result['name'],'arguments':result['arguments']}},{'role':'tool','name':result['name'],'data':data}])
    raise DirectError('STEP_LIMIT')


def run_case(case, topology, server_executable, server_script, proxy_executable, proxy_script):
    with tempfile.TemporaryDirectory(prefix='psp-topology-') as directory:
        spy, proxy_spy = str(Path(directory)/'server.jsonl'), str(Path(directory)/'proxy.jsonl')
        through_proxy = topology == 'C' and case['kind'] != 'bypass'
        def events(complete=False): return read_events(spy,case['id'],complete)
        peer = f = None
        try:
            data, p = case['input'], case['input']['perturbation']
            terms = tokenize_declaration(data['declaration']) if 'declaration' in data else []
            f = Fixture({'base':{'agents':','.join(data['agents'])},'providerCancel':p == 'cancel-before-dispatch','bypassWrite':p == 'stale-state',
                'displayDenied':p == 'display-denial','providerTraining':p == 'inference-denial','tamperedPrompt':p == 'tampered-prompt','expiredPrompt':p == 'expired-prompt',
                'responses':[{'type':'tool','name':data['requestedAgent'].removeprefix('mcp://').replace('/','.'),'arguments':{'recordId':data['recordId'],**data.get('extraArguments',{})}},{'type':'final','text':'unused'}]})
            peer = StdioMcpClient.connect({'executable':proxy_executable if through_proxy else server_executable,
                'args':[proxy_script,case['id'],server_executable,server_script,spy,proxy_spy] if through_proxy else [server_script,topology,case['id'],spy],
                'env':environment(),'serverInfo':{'name':'psp-cdl-reference','version':'0.1.0'},'timeoutMs':8000})
            names = ['allowed','other'] + (['export'] if topology == 'A' or case['kind'] == 'bypass' else [])
            registrations = peer.registrations('reference',[approval(name,data['capabilities']) for name in names],f.base.now)
            control = peer.registrations('fixture-control',[close_approval()],f.base.now)[0] if through_proxy else None
            if through_proxy and p == 'cancel-after-read':
                for r in registrations:
                    original = r['invoke']
                    r['invoke'] = lambda args,options,invoke=original:invoke(args,{**options,'cancelled':lambda:False})
            policy = f.base.policy
            def dispatch_policy(*args):
                result = policy(*args)
                result['resources'][0]['covenants'] = data['covenants']
                return result
            f.base.policy = dispatch_policy
            if data.get('kid'):
                f.prompt = lambda _p,b:sign_envelope('Use the synthetic read tool when needed.',
                    {'algorithm':'ed25519','signatureVersion':'2.0','kid':data['kid'],'timestamp':900,'expires':1700,'version':'1.0.0','sectionType':'system','contentType':'text','trustLevel':2,'attributes':prompt_context(b)},bytes([19])*32)
                f.verification = lambda *_:{'keys':data['trustedKeys']}
            invoke = f.provider['invoke']
            def provider_invoke(request, options):
                response = invoke(request,options)
                if response['type'] == 'final':
                    message = next(m for m in reversed(request['messages']) if m['role'] == 'tool')
                    return {'type':'final','text':message['data']['message']}
                return response
            provider = {**f.provider,'invoke':provider_invoke}
            gate = McpDispatchGate(f.base.store,f.base,'registry-1',[r for r in registrations if r['readOnly']])
            loop = BufferedLlmLoop(f.base.store,gate,f,provider)
            options = {**f.options,'cancelled':lambda:bool(f.base.flags.get('cancelled')) or p == 'cancel-after-read' and
                (any(e['kind'] == 'proxy-release' for e in read_events(proxy_spy,case['id'])) if through_proxy else any(e['kind'] == 'read' for e in events()))}
            codes, outputs = [], []
            message = json.dumps({'message':SUITE['configuration']['userMessage'],'agents':data['agents'],'toolCapabilities':data['capabilities'],'covenants':data['covenants'],'lexicalTerms':terms},separators=(',',':'))
            def attempt():
                try:
                    if topology == 'A': outputs.append(run_direct(provider,registrations,message,options))
                    else:
                        result = loop.run(data['token'],f.base.session['sessionId'],{'message':message},options)
                        if result['provenance']['trustLevel'] != 5 or result['provenance']['outputDigest'] != binding_digest({'text':result['text']}): raise RuntimeError('INVALID_PROVENANCE')
                        outputs.append(result['text'])
                    codes.append('OK')
                except Exception as exc:
                    if not hasattr(exc,'code'): raise
                    codes.append(exc.code)
            if case['kind'] == 'bypass':
                output = next(r for r in registrations if r['name'] == 'export')['invoke']({'recordId':data['recordId']},options)
                codes.append('OK')
                outputs.append(output['message'])
            elif p == 'replay-old-prompt':
                captured = None
                prompt = f.prompt
                def capture(*args):
                    nonlocal captured
                    captured = prompt(*args)
                    return captured
                f.prompt = capture
                attempt()
                f.prompt = lambda *_:captured
                f.base.update()
                attempt()
            else: attempt()
            state = f.base.store.execute(f.base.actor,{'action':'getSession','sessionId':f.base.session['sessionId']})
            if control and not Path(spy+'.closed').exists(): control['invoke']({},{**options,'cancelled':lambda:False})
            peer.close()
            peer = None
            if through_proxy:
                for _ in range(100):
                    if Path(spy+'.closed').exists(): break
                    time.sleep(.02)
                if not Path(spy+'.closed').exists(): raise RuntimeError('SERVER_CLEANUP_FAILED')
            observed = events(True)
            proxy_events = read_events(proxy_spy,case['id'],True) if through_proxy else []
            if observed[0]['kind'] != 'isolation-probes-blocked' or through_proxy and proxy_events[0]['kind'] != 'isolation-probes-blocked': raise RuntimeError('ISOLATION_NOT_OBSERVED')
            secrets = ['test-owner','test-tenant','tenant-a','subject-a','test-signing-key',CREDENTIAL,f.base.session['sessionId'],'sessionVersion','fixture_close']
            return {'codes':codes,'terms':terms,'providerCalls':len(f.requests),'events':observed,'proxyEvents':proxy_events,'outputs':outputs,
                'providerToolMessages':[m['data']['message'] for r in f.requests for m in r['messages'] if m['role'] == 'tool'],
                'sessionVersion':None if topology == 'A' else state['version'],'authorityLeak':any(v in json.dumps(f.requests) for v in secrets)}
        finally:
            try:
                if peer: peer.close()
            finally:
                if f: f.close()


if __name__ == '__main__':
    topology, *args = sys.argv[1:]
    report = []
    for case in SUITE['cases']:
        if case['expected'][topology] is None: continue
        try: report.append({'id':case['id'],'observation':run_case(case,topology,*args)})
        except Exception as exc:
            report.append({'id':case['id'],'error':'ADAPTER_ERROR'})
            if os.environ.get('PSP_MATRIX_DEBUG') == '1': print(case['id'],type(exc).__name__,str(exc),file=sys.stderr)
    print(json.dumps(report))

File: scripts/topology-adapter.mjs
// SPDX-License-Identifier: Apache-2.0
// Offline A/B/C execution. Expectations are consumed only by the separate grader.
import {mkdtempSync,rmSync,existsSync} from 'node:fs';
import {tmpdir} from 'node:os';
import {join} from 'node:path';
import {pathToFileURL} from 'node:url';
import {fixture} from './llm-fixtures.mjs';
import {BufferedLlmLoop,promptContext} from '@psp-cdl/llmproxy';
import {McpDispatchGate,bindingDigest} from '@psp-cdl/mcpproxy';
import {StdioMcpClient} from '@psp-cdl/mcpproxy/mcp';
import {signEnvelope} from '@psp-cdl/core/crypto';
import {tokenizeDeclaration} from '@psp-cdl/cdl';
import {suite,approval,environment,credential,readEvents,closeApproval} from './topology-common.mjs';

// A transport/orchestration adapter, with no PSP/CDL decision point. The callback
// can be replaced by a host-owned provider; CI supplies a deterministic script.
export async function runDirect(provider,registrations,message,options) {
  const check=()=>{if(options.cancelled())throw Object.assign(Error('CANCELLED'),{code:'CANCELLED'});};
  const messages=[{role:'system',content:'Use the synthetic read tool when needed.'},{role:'user',content:message}];
  const tools=registrations.map(r=>({name:r.server+'.'+r.name,inputSchema:r.inputSchema,outputSchema:r.outputSchema}));
  for(let step=0;step<options.maxSteps;step++) {
    check();const result=await provider.invoke(structuredClone({messages,tools}),options);check();
    if(!result||typeof result!=='object'||Array.isArray(result))throw Object.assign(Error('INVALID_RESPONSE'),{code:'INVALID_RESPONSE'});
    if(result.type==='final'&&typeof result.text==='string')return result.text;
    const tool=registrations.find(r=>r.server+'.'+r.name===result.name);
    if(result.type!=='tool'||!tool||!result.arguments||typeof result.arguments!=='object'||Array.isArray(result.arguments))throw Object.assign(Error('INVALID_RESPONSE'),{code:'INVALID_RESPONSE'});
    let data;
    try {data=await tool.invoke(result.arguments,options);}catch(error){check();throw error;}
    check();messages.push({role:'assistant',call:{name:result.name,arguments:result.arguments}},{role:'tool',name:result.name,data});
  }
  throw Object.assign(Error('STEP_LIMIT'),{code:'STEP_LIMIT'});
}

export async function runCase(c,topology,serverExecutable,serverScript,proxyExecutable,proxyScript) {
  const directory=mkdtempSync(join(tmpdir(),'psp-topology-')),spy=join(directory,'server.jsonl'),proxySpy=join(directory,'proxy.jsonl');
  let peer,f;
  const throughProxy=topology==='C'&&c.kind!=='bypass';
  const events=complete=>readEvents(spy,c.id,complete);
  try {
    const input=c.input,p=input.perturbation;
    const terms=Object.hasOwn(input,'declaration')?tokenizeDeclaration(input.declaration):[];
    f=await fixture({base:{agents:input.agents.join(',')},providerCancel:p==='cancel-before-dispatch',bypassWrite:p==='stale-state',
      displayDenied:p==='display-denial',providerTraining:p==='inference-denial',tamperedPrompt:p==='tampered-prompt',expiredPrompt:p==='expired-prompt',
      responses:[{type:'tool',name:input.requestedAgent.replace('mcp://','').replace('/','.'),arguments:{recordId:input.recordId,...input.extraArguments}},{type:'final',text:'unused'}]});
    peer=await StdioMcpClient.connect({executable:throughProxy?proxyExecutable:serverExecutable,
      args:throughProxy?[proxyScript,c.id,serverExecutable,serverScript,spy,proxySpy]:[serverScript,topology,c.id,spy],
      env:environment(),serverInfo:{name:'psp-cdl-reference',version:'0.1.0'},timeoutMs:8000});
    const registrations=peer.registrations('reference',['allowed','other',...(topology==='A'||c.kind==='bypass'?['export']:[])].map(name=>approval(name,input.capabilities)),f.host.now);
    const control=throughProxy?peer.registrations('fixture-control',[closeApproval()],f.host.now)[0]:null;
    if(throughProxy&&p==='cancel-after-read')for(const r of registrations) {
      const invoke=r.invoke;r.invoke=(args,options)=>invoke(args,{...options,cancelled:()=>false});
    }
    const dispatchPolicy=f.host.policy;
    f.host.policy=async(...args)=>{const result=await dispatchPolicy(...args);result.resources[0].covenants=input.covenants;return result;};
    if(input.kid) {
      f.loopHost.prompt=(_p,b)=>signEnvelope('Use the synthetic read tool when needed.',
        {algorithm:'ed25519',signatureVersion:'2.0',kid:input.kid,timestamp:900,expires:1700,version:'1.0.0',sectionType:'system',contentType:'text',trustLevel:2,attributes:promptContext(b)},new Uint8Array(32).fill(19));
      f.loopHost.verification=()=>({keys:input.trustedKeys});
    }
    const invoke=f.provider.invoke;
    const provider={...f.provider,invoke:async(request,options)=>{
      const response=await invoke(request,options);
      if(response.type==='final') {
        const message=request.messages.findLast(m=>m.role==='tool');
        if(!message)throw Error('MISSING_TOOL_OBSERVATION');
        return {type:'final',text:message.data.message};
      }
      return response;
    }};
    const gate=new McpDispatchGate(f.store,f.host,'registry-1',registrations.filter(r=>r.readOnly));
    const loop=new BufferedLlmLoop(f.store,gate,f.loopHost,provider);
    const options={...f.options,cancelled:()=>!!f.baseFlags.cancelled||(p==='cancel-after-read'&&
      (throughProxy?readEvents(proxySpy,c.id).some(e=>e.kind==='proxy-release'):events(false).some(e=>e.kind==='read')))};
    const codes=[],outputs=[];
    const message=JSON.stringify({message:suite.configuration.userMessage,agents:input.agents,toolCapabilities:input.capabilities,covenants:input.covenants,lexicalTerms:terms});
    const attempt=async()=>{
      try {
        if(topology==='A')outputs.push(await runDirect(provider,registrations,message,options));
        else {
          const result=await loop.run(input.token,f.session.sessionId,{message},options);
          if(result.provenance.trustLevel!==5||result.provenance.outputDigest!==bindingDigest({text:result.text}))throw Error('INVALID_PROVENANCE');
          outputs.push(result.text);
        }
        codes.push('OK');
      }catch(error){if(!error.code)throw error;codes.push(error.code);}
    };
    if(c.kind==='bypass') {
      const data=await registrations.find(r=>r.name==='export').invoke({recordId:input.recordId},options);
      codes.push('OK');outputs.push(data.message);
    }else if(p==='replay-old-prompt') {
      let captured;const prompt=f.loopHost.prompt;
      f.loopHost.prompt=(...args)=>(captured=prompt(...args));await attempt();
      f.loopHost.prompt=()=>captured;await f.update();await attempt();
    }else await attempt();
    const state=await f.store.execute(f.actor,{action:'getSession',sessionId:f.session.sessionId});
    if(control&&!existsSync(spy+'.closed'))await control.invoke({},{...options,cancelled:()=>false});
    await peer.close();peer=null;
    if(throughProxy) {
      for(let i=0;!existsSync(spy+'.closed')&&i<100;i++)await new Promise(r=>setTimeout(r,20));
      if(!existsSync(spy+'.closed'))throw Error('SERVER_CLEANUP_FAILED');
    }
    const observed=events(true),proxyEvents=throughProxy?readEvents(proxySpy,c.id,true):[];
    if(observed[0]?.kind!=='isolation-probes-blocked'||throughProxy&&proxyEvents[0]?.kind!=='isolation-probes-blocked')throw Error('ISOLATION_NOT_OBSERVED');
    const secrets=['test-owner','test-tenant','tenant-a','subject-a','test-signing-key',credential,f.session.sessionId,'sessionVersion','fixture_close'];
    return {codes,terms,providerCalls:f.requests.length,events:observed,proxyEvents,outputs,
      providerToolMessages:f.requests.flatMap(r=>r.messages.filter(m=>m.role==='tool').map(m=>m.data.message)),
      sessionVersion:topology==='A'?null:state.version,authorityLeak:secrets.some(value=>JSON.stringify(f.requests).includes(value))};
  }finally {try{await peer?.close();}finally{f?.close();rmSync(directory,{recursive:true,force:true,maxRetries:20,retryDelay:50});}}
}

if(process.argv[1]&&import.meta.url===pathToFileURL(process.argv[1]).href) {
  const [topology,...args]=process.argv.slice(2),report=[];
  for(const c of suite.cases.filter(c=>c.expected[topology]!==null)) {
    try {report.push({id:c.id,observation:await runCase(c,topology,...args)});}
    catch(error){report.push({id:c.id,error:'ADAPTER_ERROR'});if(process.env.PSP_MATRIX_DEBUG==='1')process.stderr.write(c.id+': '+String(error)+'\n');}
  }
  console.log(JSON.stringify(report));
}
